### PR TITLE
Support no authentication on registry

### PR DESCRIPTION
**Pull request by bschelberg**
I'm not sure if it's possible with the newer version of the Docker registry, but with the old version that we're running, we don't have authentication on our registry. This led to some obscure error messages when trying to publish Docker images to our registry from a Jenkins build. The attached pull request supports no authentication when pushing to a registry.

---

This change is [Reviewable](https://reviewable.io/reviews/docker-java/docker-java/693)

### CODE REVIEW

**Comment by KostyaSha** [comment]
So what happens when you set auth to null or didn't specify it for push?

**Comment by bschelberg** [comment]
If auth is not specified, PushImageCmdExec sends an X-Registry-Auth header with the value bnVsbA== (base 64 encoded value of "null").

I tried sending an empty String, but it also caused problems with Docker.

It might be better to return an empty String or null from registryAuth() and then check for empty String or null in PushImageCmdExec before adding the header, I'm not sure.

File: src/main/java/com/github/dockerjava/jaxrs/AbstrDockerCmdExec.java [hunk 1]
```diff
@@ -40,7 +40,8 @@ protected AuthConfigurations getBuildAuthConfigs() {
 
     protected String registryAuth(AuthConfig authConfig) {
         try {
-            return Base64.encodeBase64String(new ObjectMapper().writeValueAsString(authConfig).getBytes());
+            String json = authConfig == null ? "{}" : new ObjectMapper().writeValueAsString(authConfig);
```
Comment (KostyaSha) [inline src/main/java/com/github/dockerjava/jaxrs/AbstrDockerCmdExec.java]: Could you move `{}` to static field `EMPTY_JSON` ?

**Comment by KostyaSha** [comment]
Could you also specify old and new versions? What backend do authentication? And what error docker daemon raise?

**Comment by bschelberg** [comment]
Sure, I can add empty object as a constant.

Unfortunately, I don't have the exact version of the registry which we're using, I just pulled the latest tag. According to docker, that was 2.144840 years ago. The docker client that we're using to do the push is version 1.10.3 (Red Hat 7).

I'm not sure what you're question about backends and authentication means. The registry we're running is on a private network, quite old, and doesn't do authentication at all.

The error in the docker daemon (on the client doing the push):
```
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: time="2016-10-17T06:41:53.411739538Z" level=info msg="{Action=push, LoginUID unknown, PID unknown}"
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: 2016/10/17 06:41:53 http: panic serving 10.30.102.46:26340: assignment to entry in nil map
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: goroutine 13809 [running]:
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: net/http.(*conn).serve.func1(0xc8228e3000)
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /usr/lib/golang/src/net/http/server.go:1389 +0xc1
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: panic(0x121c3e0, 0xc8213be530)
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /usr/lib/golang/src/runtime/panic.go:443 +0x4e9
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: github.com/docker/docker/api/server/router/local.(*router).getAuthConfigs(0xc820f442a0, 0x7fb54c174b18, 0xc8213be4b0, 0xc823d24380, 0x21c0001, 0x0, 0x0, 0xc821590480, 0x0, 
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /builddir/build/BUILD/docker-58b38794e965c6f4df7b11883eb25f992ae2a627/_build/src/github.com/docker/docker/api/server/router/local/image.go:436 +0x830
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: github.com/docker/docker/api/server/router/local.(*router).postImagesPush(0xc820f442a0, 0x7fb5535822a0, 0xc8238945d0, 0x7fb553582268, 0xc820c6ad20, 0xc823d24380, 0xc82209c3
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /builddir/build/BUILD/docker-58b38794e965c6f4df7b11883eb25f992ae2a627/_build/src/github.com/docker/docker/api/server/router/local/image.go:213 +0x4e4
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: github.com/docker/docker/api/server/router/local.(*router).(github.com/docker/docker/api/server/router/local.postImagesPush)-fm(0x7fb5535822a0, 0xc8238945d0, 0x7fb553582268
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /builddir/build/BUILD/docker-58b38794e965c6f4df7b11883eb25f992ae2a627/_build/src/github.com/docker/docker/api/server/router/local/local.go:102 +0x74
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: github.com/docker/docker/api/server.(*Server).auditMiddleware.func1(0x7fb5535822a0, 0xc8238945d0, 0x7fb553582268, 0xc820c6ad20, 0xc823d24380, 0xc82209c360, 0x0, 0x0)
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /builddir/build/BUILD/docker-58b38794e965c6f4df7b11883eb25f992ae2a627/_build/src/github.com/docker/docker/api/server/middleware.go:163 +0xa2
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: github.com/docker/docker/api/server.versionMiddleware.func1(0x7fb5535822a0, 0xc8238945d0, 0x7fb553582268, 0xc820c6ad20, 0xc823d24380, 0xc82209c360, 0x0, 0x0)
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /builddir/build/BUILD/docker-58b38794e965c6f4df7b11883eb25f992ae2a627/_build/src/github.com/docker/docker/api/server/middleware.go:155 +0x83a
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: github.com/docker/docker/api/server.(*Server).corsMiddleware.func1(0x7fb55367e4c0, 0xc82000f330, 0x7fb553582268, 0xc820c6ad20, 0xc823d24380, 0xc82209c360, 0x0, 0x0)
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /builddir/build/BUILD/docker-58b38794e965c6f4df7b11883eb25f992ae2a627/_build/src/github.com/docker/docker/api/server/middleware.go:134 +0xfa
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: github.com/docker/docker/api/server.(*Server).userAgentMiddleware.func1(0x7fb55367e4c0, 0xc82000f330, 0x7fb553582268, 0xc820c6ad20, 0xc823d24380, 0xc82209c360, 0x0, 0x0)
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /builddir/build/BUILD/docker-58b38794e965c6f4df7b11883eb25f992ae2a627/_build/src/github.com/docker/docker/api/server/middleware.go:117 +0x4af
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: github.com/docker/docker/api/server.(*Server).authorizationMiddleware.func1(0x7fb55367e4c0, 0xc82000f330, 0x7fb5535fecb8, 0xc823d81930, 0xc823d24380, 0xc82209c360, 0x0, 0x0
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /builddir/build/BUILD/docker-58b38794e965c6f4df7b11883eb25f992ae2a627/_build/src/github.com/docker/docker/api/server/middleware.go:86 +0x5ba
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: github.com/docker/docker/api/server.(*Server).makeHTTPHandler.func1(0x7fb5535fecb8, 0xc823d81930, 0xc823d24380)
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /builddir/build/BUILD/docker-58b38794e965c6f4df7b11883eb25f992ae2a627/_build/src/github.com/docker/docker/api/server/server.go:175 +0x2ad
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: net/http.HandlerFunc.ServeHTTP(0xc8208ff640, 0x7fb5535fecb8, 0xc823d81930, 0xc823d24380)
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /usr/lib/golang/src/net/http/server.go:1618 +0x3a
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: github.com/gorilla/mux.(*Router).ServeHTTP(0xc820fe0000, 0x7fb5535fecb8, 0xc823d81930, 0xc823d24380)
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /builddir/build/BUILD/docker-58b38794e965c6f4df7b11883eb25f992ae2a627/vendor/src/github.com/gorilla/mux/mux.go:98 +0x29e
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: github.com/docker/docker/api/server.(*routerSwapper).ServeHTTP(0xc820c38ac0, 0x7fb5535fecb8, 0xc823d81930, 0xc823d24380)
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /builddir/build/BUILD/docker-58b38794e965c6f4df7b11883eb25f992ae2a627/_build/src/github.com/docker/docker/api/server/router_swapper.go:29 +0x72
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: net/http.serverHandler.ServeHTTP(0xc8201c1180, 0x7fb5535fecb8, 0xc823d81930, 0xc823d24380)
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /usr/lib/golang/src/net/http/server.go:2081 +0x19e
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: net/http.(*conn).serve(0xc8228e3000)
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /usr/lib/golang/src/net/http/server.go:1472 +0xf2e
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: created by net/http.(*Server).Serve
Oct 17 06:41:53 ci-agent10.dev.arsrs.local docker-current[16178]: /usr/lib/golang/src/net/http/server.go:2137 +0x44e
```

**Comment by KostyaSha** [comment]
Is it still actual? Could you rebase and fixup/squash commits? I remember we fixed something about empty jsons for netty or jersey...

**Comment by bschelberg** [comment]
Seems to work properly now with version 0.16.2.

**Pull request closed without merging**